Photon: PANIC when TCPClient fails to connect

When trying to connect to a closed port or just failing to connect, the Photon restarts with an SOS instead of simply returning false.

Update:
- crashes and restarts w a PANIC if you try to connect right after the unit has connected to the internet (if you wait 2 seconds it works, less and it fails)
- crashes and restart w a PANIC if: address doesn't exist, nobody is listening on the port or if the tcp connection fails after a successful connect

Build fails for me:

...
Building file: src/hal_peripherals.c
Invoking: ARM GCC C Compiler
mkdir -p ../build/target/hal-dynalib/platform-6-m/./src/
arm-none-eabi-gcc -DSTM32_DEVICE -DSTM32F2XX -DPLATFORM_THREADING=1 -DPLATFORM_ID=6 -DPLATFORM_NAME=photon -DUSBD_VID_SPARK=0x2B04 -DUSBD_PID_DFU=0xD006 -DUSBD_PID_CDC=0xC006 -DDYNALIB_IMPORT -DUSER_FIRMWARE_IMAGE_SIZE=0x20000 -DUSER_FIRMWARE_IMAGE_LOCATION=0x80A0000 -DMODULAR_FIRMWARE=1 -DMODULE_VERSION=4 -DMODULE_FUNCTION=4 -DMODULE_INDEX=1 -DMODULE_DEPENDENCY=0,0,0 -DRELEASE_BUILD -Werror -I../hal/inc -I../hal/shared -I../hal/src/photon/api -I../dynalib/inc -I. -MD -MP -MF ../build/target/hal-dynalib/platform-6-m/./src/hal_peripherals.o.d -ffunction-sections -fdata-sections -Wall -Wno-switch -Wno-error=deprecated-declarations -fmessage-length=0 -fno-strict-aliasing -DSPARK=1 -DSTART_DFU_FLASHER_SERIAL_SPEED=14400 -DSTART_YMODEM_FLASHER_SERIAL_SPEED=28800 -g3 -gdwarf-2 -Os -mcpu=cortex-m3 -mthumb -Wno-pointer-sign -std=gnu99  -c -o ../build/target/hal-dynalib/platform-6-m/./src/hal_peripherals.o src/hal_peripherals.c
In file included from src/hal_peripherals.c:1:0:
../hal/inc/hal_dynalib_peripherals.h:32:56: fatal error: platform_config.h: No such file or directory
 #include "platform_config.h"    // for HAS_SERIAL_FLASH
                                                        ^
compilation terminated.
make[2]: *** [../build/target/hal-dynalib/platform-6-m/./src/hal_peripherals.o] Error 1
make[2]: Leaving directory `/home/andyw/spark/firmware/hal-dynalib'
make[1]: *** [hal-dynalib] Error 2
make[1]: Leaving directory `/home/andyw/spark/firmware/modules/photon/system-part1'
make: *** [/home/andyw/spark/firmware/modules/photon/system-part1/makefile] Error 2

--- a/hal/src/photon/socket_hal.cpp
+++ b/hal/src/photon/socket_hal.cpp
@@ -107,20 +107,14 @@ struct tcp_packet_t
  */
 struct tcp_socket_t : public wiced_tcp_socket_t {
     tcp_packet_t packet;
-    bool open; bool opened;
+    bool open;
     volatile bool closed_externally;
 
-    tcp_socket_t() : open(false), opened(false), closed_externally(false) {}
+    tcp_socket_t() : open(false), closed_externally(false) {}
 
-    ~tcp_socket_t()
-    {
-        // calling wiced_tcp_delete_socket when the socket didn't connect causes a hard fault.
-        // connected is only called after successful connection, setting opened to true.
-        if (opened)
-            wiced_tcp_delete_socket(this);
-    }
+    ~tcp_socket_t() { wiced_tcp_delete_socket(this); }
 
-    void connected() { open = true; opened = true; }
+    void connected() { open = true; }
 
     bool isClosed() {
         return !open || closed_externally;
@@ -630,8 +624,14 @@ sock_result_t socket_connect(sock_handle_t sd, const sockaddr_t *addr, long addr
             SOCKADDR_TO_PORT_AND_IPADDR(addr, addr_data, port, ip_addr);
             unsigned timeout = 5*1000;
             result = wiced_tcp_connect(tcp_socket, &ip_addr, port, timeout);
-            if (result==WICED_SUCCESS)
+            if (result==WICED_SUCCESS) {
                 tcp_socket->connected();
+            } else {
+              // Work around WICED bug that doesn't set connection handler to NULL after deleting
+              // it, leading to deleting the same memory twice and a crash
+              // WICED/network/LwIP/WICED/tcpip.c:920
+              tcp_socket->conn_handler = NULL;
+            }
         }
     }
     return as_sock_result(result);

